### ISSUE TITLE
better advertise error tracking to the user

### ISSUE THREAD

**Issue by Ant1r** [pure-data/pure-data#1508]
Error tracking (i.e: ctrl-clicking an error or message posted to the Pd console, allowing to open the window and select the object from which the error/message originated) is a great Pd feature. I'm even wondering if the `Edit/Find last error` is still really needed (personally I quite never use it).

I've spent some time last year(s) to make sure most errors/messages (even from `[print]`, `[messages(`, GOP subpatches and  `[clones]`) are ctrl-clickable.

Unfortunately, I suspect many Pd users (even experimented ones) are not aware of it. Actually, it seems it's documented neither advertised nowhere...

So I think we
- should agree on the more accurate and concise naming for this function (is "error tracking" the best name?)
- document it where applicable (Pd manual and other places)
- find a way to make it obvious to anyone, maybe by posting a message (in red?) a the beginning of the Pd window.

What do you think?

**Comment by Spacechild1** [comment]
Yes! It's definitely not obvious that you can track errors with Ctrl+click.

> find a way to make it obvious to anyone, maybe by posting a message (in red?) a the beginning of the Pd window.

Maybe post such a message with the very first error message?

**Comment by Ant1r** [comment]
> the very first error message

yes, but this will not trigger for other traceable messages (e.g from [print].

Actually, there's currently such a message, but 
- it only talks about `Menu/Find Last Error`
- it's only triggered by `pd_error()`
- it's posted to log level 4, so most of the times you can't read it...

**Comment by Ant1r** [comment]
What would you think of something like that:
![Capture du 2021-12-07 11-19-26](https://user-images.githubusercontent.com/3748246/145011338-565bdbd1-065f-482b-bcc9-ba91037c52d9.png)
Rewording welcome ;-)

**Comment by Ant1r** [comment]
I was also thinking that maybe double-click would be better than ctrl-click (we also can have both)?
Even if it collides a bit with text selecting.

**Comment by umlaeute** [comment]
> I was also thinking that maybe double-click would be better than ctrl-click (we also can have both)?

why?

> Even if it collides a bit with text selecting.

where is the *bit* that is not colliding :-P ?

probably a right-click context menu would be even better?

**Comment by umlaeute** [comment]
> Ctrl-click any message or error to locate its origin, when possible

how about:
> You can Ctrl-click most lines in the Pd-console to locate the object that emitted it.

in general i think it would be better to have a "Tip of the Day" popup (that can be disabled :-)) with such information (rather than putting these hints on places everywhere - you really only need to learn this trick *once* and afterwards the printout becomes just noise)

**Comment by Ant1r** [comment]
> > double-click would be better than ctrl-click (we also can have both)?
> 
> why?

because you can do it with only one hand? and maybe a more common convention? But, I'm just asking!
 

> where is the _bit_ that is not colliding :-P ?

well... there is none :-) but at least it doesn't prevent to select the text.

Anyway I'm all OK to keep ctrl-click.
Right-click context menu seems unnecessary to me: I don't see what other entry you would put in the menu, and the gesture would be a little more complex.

BTW, currently each message is also bound to "Return" and "KP_Enter" keys. I'm not sure it's really used by anyone, and I guess this has some impact on the memory usage. Maybe we could remove these shortcuts?


> "Tip of the Day"

mmm. why not, even I'm not sure it'll please everyone! 
Anyway it would need some more 'tips' than just this one to justify its existence (and the related work time..).
But I understand your objection about reading the message each time you start Pd while you're already aware of it.

**Comment by Ant1r** [comment]
The message could add: `(you can Ctrl-click this message if you don't want to see it anymore)`.
Ctrl-clicking the message would add a flag into the guiprefs that would prevent to print it at next start.

To be complete, a button should be added somewhere in the preferences dialogs. Maybe it could be good to build such a solution starting from #1397?

**Comment by umlaeute** [comment]
to be honest, i'd rather not have thousands of miniature options to turn on/off single hints.

i like the tip-of-the-day option better by the minute (and yes: i'm always annoyed by tip-of-the-day dialogs myself; probably it is just enough to have such a tip-of-the-day *disabled* by default - but you can invoke it from the help-menu)

**Comment by Spacechild1** [comment]
Personally, I would prefer to have a message like in https://github.com/pure-data/pure-data/issues/1508#issuecomment-987779005, but only for the very first error.

In fact, `pd_error` already almost does something like that:
```
if (object && !saidit)
    {
        if (sys_havegui())
            logpost(NULL, 4,
                "... you might be able to track this down from the Find menu.");
        saidit = 1;
    }
```
But the message does not mention "Ctrl/Cmd + Click" and is only shown at "Verbose". I think this really should be printed at "Normal".

**Comment by Ant1r** [comment]
What if we just get rid of the modifier key?

I mean, obvious things don't have to be explained, and I guess lots of users try to click the error, but are discouraged because nothing happens.

I've just done a little test, and it is not that annoying when selecting text; moreover it could be improved with a position-delta threshold: if you move the mouse between mouse-down and mouse-up then the "Find error" is not called.

**Comment by sebshader** [comment]
+1 for keeping modifier-click by default. regular-click could be annoying if you accidentally click on an error in the console (e.g. when trying to bring it into focus/raise it)
I also don't think regular-clicking is something people necessarily assume will bring them to the error.. in fact I think it's more reasonable for people to try to copy-paste.
Users can always change it via tcl plugin I think, but such 'special' functionality is more appropriate for a modifier key imo.

**Comment by Spacechild1** [comment]
Yes, simple clicking would probably be most intuitive. After all, people already manage to select text that contains hyperlinks :) I guess I agree that we should get rid of the modifier.

**Comment by Spacechild1** [comment]
> regular-click could be annoying if you accidentally click on an error in the console (e.g. when trying to bring it into focus/raise it)

That's a good point.

IMO a post in the console would be the most helpful thing. Then I don't care too much if it's single click, double click or with some modifier.

**Comment by porres** [comment]
just mentioning this is on my list on things to add in the manual

I also agree we could improve how it works and is presented to the user

**Comment by porres** [comment]
> just mentioning this is on my list on things to add in the manual

I have added this to the manual, is it good enough? can we close this?

**Comment by php4fan** [comment]
OMG I would NEVER have found it!

I've been "away" from Pd for like 13 years, I'm digging up some old patches now, and I was like "I can't believe that after all these years we can still only track the last error", then found this issue by pure chance.


> I have added this to the manual, is it good enough? can we close this?

Nooooooo, something like this should be discoverable.

This seems the best proposal to me:
https://github.com/pure-data/pure-data/issues/1508#issuecomment-987779005
super simple and I see no downsides.

The "tip of the day" would solve nothing: very few people read the tip of the day. 

Simple click and double click are no-no's, they interfere with text selection. Ctrl+click is fine, the problem is making the feature discoverable.

**Right-click->"find source"** would be great too. 

> Right-click context menu seems unnecessary to me: I don't see what other entry you would put in the menu, and the gesture would be a little more complex.

Even if it's the only entry. Of course it wouldn't _replace_ the ctrl+click. Its purpose is to make the feature discoverable. People are used to right-clicking on things to see if you can do something with them, it's a very natural way to discover a feature. The context menu entry could read "Find source (Ctrl+click)" so it also tells you about the ctrl+click option.

One problem is that plain text is not something that you would ever dream of being interactive, so even the right-click can be missed.

A "▶" or a bullet or somethig on the left before the text of the error would help suggesting that there's an interactive element there, and may increase the chances that one tries right-clicking. Also, just left-clicking the "▶"/bullet (NOT the rest of the text) could result in opening the context menu (with the "Find source (Ctrl+click)" option) as if you had right-clicked on any other part of the text. (or it could just activate the feature directly, but I don't like that, as you may still never discover the ctrl+click).

Another place where the general hint could be added (not as an alternative to the above, but as an addition) is when you use "Find last error", unless/until that gets removed completely.

**Comment by php4fan** [comment]
![Image](https://github.com/user-attachments/assets/37c1e27b-5e09-4e44-9160-ef22161fabfc)

BTW is this weird cursor supposed to be a way of suggesting that the text is ctrl+clickable, or is it just an unrelated bug?

**Comment by umlaeute** [comment]
> BTW is this weird cursor supposed to be a way of suggesting that the text is ctrl+clickable, or is it just an unrelated bug?

afacit it is intended to show that the error is somehow clickable.
https://github.com/pure-data/pure-data/blob/eaaf7991748a71412686a6d4a52972d67c4d4933/tcl/pdwindow.tcl#L413-L415

i have no idea what exactly the `'findinstance' mode` is. the code is rather old...

**Comment by ben-wes** [comment]
+1 for (additional) context menu on right click.

some people on macOS are also used to opening a context menu via `ctrl`+`click`. in that case, it might be logical to handle this similar to a right click and only use the `cmd`+`click` for the "Find source" function (like the temporary run mode in a patch canvas).

besides the "Find source" entry in a context menu, could there also be a "Find external" option if the object wasn't found? or is there no way to clearly distinguish this from cases where the creation failed for other reasons?

**Comment by umlaeute** [comment]
> besides the "Find source" entry in a context menu, could there also be a "Find external" option if the object wasn't found? or is there no way to clearly distinguish this from cases where the creation failed for other reasons?

if the object wasn't found, we still get the automatic "Find external" when clicking the error.
making this optional, would require changing the over-the-wire format for the Pd-GUI communication, which I'm somewhat hesitant to do.
